### DISCUSSION TITLE
Citations in custom RAG pipeline

### DISCUSSION THREAD

**Discussion by ClancyD-SHC** [open-webui/pipelines#156] [Q&A]
Can we pass through RAG metadata to show as chunks like with the local RAG interface from a pipeline?

Any help appreciated!

**Comment by csvance** [comment]
The biggest problem with OpenWebUI / Pipelines for custom RAG applications is not supporting citations. There are cases even with 400B+ parameter models where the LLM does not correctly synthesize the information from the retrieved documents or understand that it does not have sufficient context to answer. In these cases, the user has no easy way to verify the information since pipelines can only return a string (or character generator for streaming output). 

RAG without citations is dangerous. I'm open to trying to contribute something here, but the issue is with current pipeline abstractions there is no way to handle this. 

The issue with including all of the context in the response string is this will quickly saturate the context length for the chat history. Ideally the history shouldn't contain all of the context information for every previous response, it should only contain the users messages and the agents response. Even if the context was included in a way that was filtered out of the history, its still not a great way to present the information to the user.

A potential workaround creating an HTTP endpoint on the custom RAG service to display all of the context, and including a link to it in a way that would be parsed out for the chat history, like `[<citations>](https://custom-rag-service/citations/123)` - This is still kind of ugly and requires extra work, but that would also add potential for user provided feedback for the reranking model which is something that even OpenWebUI native RAG support lacks.

Another potential option is defining a standard XML/JSON citation schema for use at the end of the response which OpenWebUI could parse into citations. The disadvantage of this is it only works with simple pipeline structures. Both sides of the pipeline would have to be aware of the format, which would make any kind of composition or scaffolding difficult. To me most of the power of pipelines to begin with is the ability to easily connect an entire application into OpenWebUI's interface. Only a single pipeline is needed for such cases realistically because all of the heavy lifting will happen in the applications code rather than OpenWebUI's pipeline.

**Reply by antonio-lamberti** [reply]
Facing same issue here. I'm not satisfied with Openwebui's internal RAG, so I wanted to connect it to an external n8n instance that does proper RAG. However, without citations, it's pretty useless. 
Did anyone find a solution?

**Reply by nkostop** [reply]
Is there any news about this? Is it possible or is anybody working on this that we can help?

**Reply by ClancyD-SHC** [reply]
We sidestepped this issue by modifying the openwebUI html sanitizer to let the /detail/> and </summary/> tags through so the citations appear as drop down boxes. We still have to manage what goes into the context in the pipeline but it was a quick and easy way to have the citations viewable. I think someone should open a ticket to ask for the <citations> tags to be extracted and displayed like internal tag citations.

**Reply by nkostop** [reply]
This sounds good for a temporary solution!

Is there an example for this?
Is it very difficult for example to use event emitters or something like that?

**Reply by etiennebonnafoux** [reply]
Has someone open the ticket for this issue ?

**Reply by nkostop** [reply]
Is this issue related? can we add citation there maybe? 
https://github.com/open-webui/pipelines/issues/225

**Reply by etiennebonnafoux** [reply]
There is also https://github.com/open-webui/pipelines/issues/229

**Reply by Maleya** [reply]
Any news or developments on this topic?

**Reply by codefox1387** [reply]
For Pipes not Pipelines...If you decompose the existing RAG codebase in OWUI you will find that it leverages an event emitter with the type "citation." If you use this event emitter in your tool, function, pipeline you can pass data to it to generate citations in the standard OWUI format with the chips at the bottom of the response.

This won't help for pipelines though.

You will need a method to get those citations from whatever RAG service you are using to pass them to the emitter and not use them in the response body.

Here is a snippet of an example leveraging the BBC news feed example code. This example runs for every article found/used in the search and generates a citation with the source link and the content from the article:

## Emit citation for this article
                await __event_emitter__(
                    {
                        "type": "citation",
                        "data": {
                            "document": [article["content"]],
                            "metadata": [{"source": article["link"]}],
                            "source": {"name": article["title"]},
                        },
                    }
                )

**Reply by Maleya** [reply]
Hello! As far as I have understood, pipelines do not support event_emitters (see https://github.com/open-webui/pipelines/issues/225). Have you found a way around this or does this only apply for functions?

**Reply by codefox1387** [reply]
Ah you're right, pipe/pipes support OWUI internal functions but not pipelines. Sorry the terminology is a bit confusing, I edited my response.

**Comment by marchellodev** [comment]
While not ideal, I am outputting custom messages (tool calling, citations) like so:

```md
<details>
<summary>Citations</summary>
- Citation 1
- Citation 2
</details>
```

**Reply by marchellodev** [reply]
Having said that, you then have to manually pre-process messages to remove this kind of "UI" -- just to not confuse the LLM

Using `__event_emitter__` through SSE  would be much better

**Comment by pfedasullivan** [comment]
I've been able to get native citation rendering working with this method in our rag pipeline:
`
       
       async def outlet(self, body: dict, user: dict) -> dict:
        body["messages"][-1]["sources"] = self.build_citations()
        return body`

build_citations just returns an array of dictionary objects with the document, metadata, and source values like people have mentioned above.

**Reply by Maleya** [reply]
Hey @pfedasullivan, thanks for writing. I am curious, was this achieved with [pipelines](https://github.com/open-webui/pipelines), or with functions?

**Reply by pfedasullivan** [reply]
It is using pipelines. I started out with the[ pipeline example](https://github.com/open-webui/pipelines/blob/main/examples/scaffolds/example_pipeline_scaffold.py).

**Reply by marchellodev** [reply]
Is it possible to pass "sources" with OpenAI-style SSE streaming?

**Reply by Maleya** [reply]
Not the answer you are looking for perhaps: but i got citations with the `<details>` block working with  a streaming response by first exhausting the stream (generator) then yeilding the details block. I will try to do the same with the above method and report back.

**Reply by Maleya** [reply]
I seem unbale recreate the citations in a pipeline. I have a more or less placeholder pipeline with the following: 
```python
    def build_citations(self):
        return [
            {
                "type": "citation",
                "data": {
                    "document": ["This is a dummy document content."],
                    "metadata": [{"source": "https://dummy.source.link"}],
                    "source": {"name": "Dummy Source Title"},
                },
            }
        ]

    async def outlet(self, body: dict, user: dict) -> dict:
        body["messages"][-1]["sources"] = self.build_citations()
        return body
```

Am missing something important here @pfedasullivan? 
edit: sorry wrong @

**Reply by pfedasullivan** [reply]
Assuming you're running some version of open-webui 0.5.x, then I think you should just delete the key "type" along with it's value, and "data", but keep that value. Or in other words:

```
return [
            {
                "document": ["This is a dummy document content."],
                "metadata": [{"source": "https://dummy.source.link"}],
                "source": {"name": "Dummy Source Title"},
            }
        ]
```

**Reply by rhajou** [reply]
@pfedasullivan @Maleya  Hello, I am trying to add the citations to the open webui, I am having a custom RAG pipeline through fastAPI and connecting this to OpenWebui.

Can you tell me what type of pipeline should I attach to Openwebui to have the citations working? and Additionally should I do any modification in my backend (FastAPI) so I can send somehow the citations to this pipeline?

**Reply by pfedasullivan** [reply]
A pipeline or manifold pipeline should do the trick, and I was able to get citations working just in that file.

**Reply by rhajou** [reply]
Great thanks @pfedasullivan , I am adding my documents as part of my OpenAI response under response['docs'] = Dict[str, Any] , but it seems that the output of the function `pipe` is not the same as the input of `outlet` . Did I miss something?

**Reply by pfedasullivan** [reply]
Correct, check out the [example pipeline](https://github.com/open-webui/pipelines/blob/main/examples/scaffolds/example_pipeline_scaffold.py).
Here is pipe's definition: def pipe(self, user_message: str, model_id: str, messages: List[dict], body: dict) -> Union[str, Generator, Iterator]:
As you can see it outputs either a string, Generator, or Iterator.
Outlet is:  async def outlet(self, body: dict, user: dict) -> dict:


According to my understanding try avoiding thinking of it as pipe that flows into outlet. Rather, think of it as pipe runs, then outlet runs and they both are talking to open-webui.

**Reply by LuyiTian** [reply]
@pfedasullivan Hi thank you for the trick. I tried the dummy document example it seems to work. I have two questions. 1) how you add number in the citation. I can just see a clickable citation chip without No.  2) how does outlet function communite with the pipe function then, if the output of `pipe` is not the input of `outlet`, considering the RAG case where the retrieval happens inside the `pipe` and the source document generated inside `pipe` function, how does that list of source citations goes to the `outlet` ?

**Reply by pfedasullivan** [reply]
You're welcome @LuyiTian. I don't know the answer to your first question, but plan on implementing it in the future. For your second question, it doesn't, as far I can tell. They are both async methods so they are returning data back to the client independently. You can save the rag results in a shared variable between the two though. I hope that helps.

**Reply by LuyiTian** [reply]
@pfedasullivan Hi thanks for answering. actually I have taken some time to figure them out my self. for the first question. you just need to make sure you have the relevance score in your citation dict. they dont need to be real, can be all zeros. the frontend just check if they exist and render the numbers. for the second question, I think it is a bit complicated. I think the output response will appear in the input body of outlet, the outlet is indeed executed after the pipe function. however the output of pipe function seem to be stricted by pydantic, so you cannot easily add another return value like a citation dict. so yes in the end what I have come around as a solution is to create a shared variable in the class. The last issue I have is that the pipeline system is kinda outside the openwebui, so it seems I cannot do inline clickable citations.  there are no documents around the citation thing so I copy&pasted the citation frontend code to LLM to figure things out and they are helpful: https://github.com/open-webui/open-webui/blob/6fedd72e3973e1d13c9daf540350cd822826bf27/src/lib/components/chat/Messages/Citations.svelte#L4

**Reply by LuyiTian** [reply]
apart from the pipeline, I think the native design of citation system in OpenWebUI is kinda sub-optimal, especially when you have mixed source from document and from web search. the numbering break frequently and the inline citation often not in right places or not showing correctly

**Reply by ericsua** [reply]
@LuyiTian Hi, could you please provide an example of a citation dict with the relevance score as an output of the outlet function, please?

**Reply by rhajou** [reply]
@ericsua  After manual investigation, it seems if I send back to openwebui in the ChatCompletion response a key "sources" of type List[Dict] it works. 

here's an example: 
```
sources = [  {'document': [['path1.xlsx'], ['path2.xlsx']], 
                     'metadata': [{"source": "metadata1"},{"source": "metadata2"}],
                     'distances': [0.88,0.77],
                     'source': {"name": "metadata1"}
}]
```
You need to enable citations in your openwebui. 

I had to test multiple times to arrive to a coherent display of document metadata, distances and source. 
Note that document, metadata (keys: sources), distances and source (keys: name, url) are something waiting from openwebui so it can show the citations.

**Reply by Maleya** [reply]
@ericsua You need to pass a dict element `distances` of type `list[float]`. I found this by digging around in the source code. I would be curious to hear if anyone found any other dict keys that are in some way loaded. Thus far I use: 

```python
        test_citations = [
            {
                "document": ["This is a dummy document content 1"],
                "metadata": [{"source": "https://default.link1"}],
                "source": {"name": "madeupnames 1"},
                "distances": [0.5],  # Relevance score 
            },
            {
                "document": ["This is a dummy document content 2"],
                "metadata": [{"source": "https://default.link2"}],
                "source": {"name": "madeupnames 2"},
            },
        ]
```

There appears to be a  `id` key present in the metadata, but I have not managed to manipulate it in a meaningful way.

**Reply by rhajou** [reply]
@Maleya I found that you can pass url as well in source, and you can aggregated mulitple documents in the list provided to "document" it helps aggregate citations in the GUI

**Reply by Maleya** [reply]
I noticed that as well! Documents with the same source are aggregated under the same reference 'pill'. Thanks though @rhajou, that one is easy to miss.

**Reply by rhajou** [reply]
@Maleya  but still I didn't understand the relationship or role of metadata (source inside) and "source" key . For the time being I am keeping both with same value... Would love a clear full documentation on citations  from the Open Webui team.

**Reply by ericsua** [reply]
@LuyiTian @Maleya @rhajou Sorry for the OT, but does anyone know if Pipelines can also return the message's stats as in the models directly handled by Open WebUI (see image below) in the metadata or in another field of the body with an example, or if there is a way to tell Open WebUI to compute them? Thanks ☺️
![screen-2025-02-24-11 42 27](https://github.com/user-attachments/assets/76c14684-18d7-40b7-aca9-e8ac13467f87)

**Reply by pfedasullivan** [reply]
@LuyiTian Oh cool. Thanks for sharing, and I learned something new. I was able to add distances using code like @Maleya mentions.

**Comment by aniforatos** [comment]
Does anyone know how to build the URL within the citation so that I can click it and have it bring me to the page number referenced in the source_nodes within the response.

**Comment by aaronzr** [comment]
I found some success with this: #229
